docs/kathara.1.ronn
m4_changequote()
kathara(1) -- A network emulation tool
=============================================

## SYNOPSIS

`kathara` [`-h`] [`-v`] <command> [<args>]  

## DESCRIPTION

Kathara is a network emulation tool based on Docker containers. The tool can be really helpful in showing interactive demos/lessons, testing production networks in a sandbox environment, or developing new network protocols.

Kathara is the spiritual successor of the notorious [Netkit](https://github.com/maxonthegit/netkit-core), hence it is cross-compatible, and inherits its language and features.

Within the Kathara environment each network device, called device, is implemented by a container, and each interconnection link is emulated by using a virtual network.

Each device can be configured to have an arbitrary number of (virtual) network interfaces.

By default, devices use a Docker Image which includes network oriented software such as routing daemons (RIP, OSPF, etc.), an HTTP server, firewalling utilities (`iptables`(8)), and diagnostic tools (`ping`(1), `traceroute`(1), `tcpdump`(1), etc.).  
By configuring the appropriate software, it is possible to faithfully emulate a specific network device (e.g., a router).  

Kathara provides two alternative interfaces to start and configure devices. A set of `v`-prefixed commands (vstart, vclean, vconfig), that allow to start and manage single devices while providing finegrained control on their configuration; and a set of `l`-prefixed commands (lstart, lclean, linfo, lrestart, lconfig), that ease setting up preconfigured network scenarios consisting of several devices.

Kathara also provides a set of global commands (connect, info, wipe, settings, check).

## KATHARA COMMANDS

The possible kathara commands are:

* `kathara-vstart`(1):
    Start a new Kathara device

* `kathara-vclean`(1):
    Stop a single Kathara device

* `kathara-vconfig`(1):
    Manage the network interfaces of a running Kathara device

* `kathara-lstart`(1):
    Start a Kathara network scenario

* `kathara-lclean`(1):
    Stop a Kathara network scenario

* `kathara-linfo`(1):
    Show information about a Kathara network scenario

* `kathara-lrestart`(1):
    Restart a Kathara network scenario

* `kathara-lconfig`(1):
    Manage the network interfaces of a running Kathara device in a network scenario

* `kathara-connect`(1):
    Connect to a Kathara device

* `kathara-exec`(1):
    Execute a command in a Kathara device

* `kathara-switch`(1):
    Manage a Kathara collision domain deployed as a managed switch

* `kathara-wipe`(1):
    Delete all Kathara devices and collision domains, optionally also delete settings

* `kathara-list`(1):
    Show all running Kathara devices

* `kathara-settings`(1):
    Show and edit settings

* `kathara-check`(1):
    Check your system environment

## FILES

Kathara stores its default configuration settings inside a file named `kathara.conf`. This file is placed in the `~/.config` folder. If it does not exist when Kathara is launched, it will be created with default settings. See `kathara.conf`(5) for information about the location of this file and for a description of its format.

m4_include(footer.txt)

## SEE ALSO

`kathara-lab.conf`(5), `kathara-lab.dep`(5), `kathara-lab.ext`(5), `kathara.conf`(5), [Kathara official site](https://www.kathara.org)


debian/changelog
kathara (0.47-1) UNRELEASED; urgency=medium

  * Re-written kathara to be Object Oriented
  * Initial release.

 -- Lorenzo <lorenzoariemma@gmail.com>  Fri, 27 Sep 2019 18:32:59 +0200


src/kathara.py
#!/usr/bin/env python3

import argparse
import logging
import multiprocessing
import os
import sys

from rich.logging import RichHandler

from Kathara import utils
from Kathara.auth.PrivilegeHandler import PrivilegeHandler
from Kathara.cli.ui.event.register import register_cli_events, unregister_cli_events
from Kathara.exceptions import SettingsError, DockerDaemonConnectionError, ClassNotFoundError, SettingsNotFoundError
from Kathara.foundation.cli.command.CommandFactory import CommandFactory
from Kathara.setting.Setting import Setting
from Kathara.strings import formatted_strings
from Kathara.version import CURRENT_VERSION

description_msg = """kathara [-h] [-v] <command> [<args>]

Possible Kathara commands are:\n
%s
""" % formatted_strings()


class KatharaEntryPoint(object):
    def __init__(self) -> None:
        parser = argparse.ArgumentParser(
            description='A network emulation tool.',
            usage=description_msg,
            add_help=False
        )

        parser.add_argument(
            '-h', '--help',
            action='help',
            default=argparse.SUPPRESS,
            help='Show an help message and exit.'
        )

        parser.add_argument(
            'command',
            help='Command to run.',
            nargs="?"
        )

        parser.add_argument(
            "-v", "--version",
            action="store_true",
            help='Print the current Kathara version.',
            required=False
        )

        # parse_args defaults to [1:] for args, but you need to
        # exclude the rest of the args too, or validation will fail
        args = parser.parse_args(sys.argv[1:2])

        if args.version:
            print('Current version: %s' % CURRENT_VERSION)
            unregister_cli_events()
            sys.exit(0)

        if args.command is None or not args.command.islower():
            parser.print_help()
            unregister_cli_events()
            sys.exit(1)

        try:
            # Check settings only if the user is not executing "settings" command.
            if "settings" not in args.command:
                Setting.get_instance().check()
        except (SettingsError, DockerDaemonConnectionError) as e:
            logging.critical(f"({type(e).__name__}) {str(e)}")
            unregister_cli_events()
            sys.exit(1)

        try:
            command_object = CommandFactory().create_instance(class_args=(args.command.capitalize(),))
        except ClassNotFoundError:
            logging.error(f"Unrecognized command `{args.command}`.")
            parser.print_help()
            unregister_cli_events()
            sys.exit(1)
        except ImportError as e:
            logging.critical(f"({type(e).__name__}) `{e.name}` is not installed in your system")
            unregister_cli_events()
            sys.exit(1)

        try:
            current_path = os.getcwd()
            exit_code = command_object.run(current_path, sys.argv[2:])
            unregister_cli_events()

            sys.exit(exit_code)
        except KeyboardInterrupt:
            if args.command not in ['exec', 'linfo', 'list', 'settings', 'save', 'switch']:
                logging.warning("You interrupted Kathara during a command. The system may be in an inconsistent "
                                "state! If you encounter any problem please run `kathara wipe`.")
            unregister_cli_events()
            sys.exit(0)
        except Exception as e:
            if Setting.get_instance().debug_level == "EXCEPTION":
                logging.exception(f"({type(e).__name__}) {str(e)}")
            else:
                logging.critical(f"({type(e).__name__}) {str(e)}")
            unregister_cli_events()
            sys.exit(1)


if __name__ == '__main__':
    try:
        Setting.get_instance().load_from_disk()
    except SettingsNotFoundError:
        Setting.get_instance().save_to_disk()

    try:
        debug_level = Setting.get_instance().debug_level
        debug_level = debug_level if debug_level != "EXCEPTION" else "DEBUG"
    except SettingsError:
        debug_level = "DEBUG"

    logging.basicConfig(
        level=debug_level, format="%(message)s",
        handlers=[RichHandler(rich_tracebacks=True, show_time=False, show_path=False)]
    )

    multiprocessing.freeze_support()

    register_cli_events()

    utils.check_python_version()

    utils.exec_by_platform(PrivilegeHandler.get_instance().drop_effective_privileges, lambda: None, lambda: None)

    KatharaEntryPoint()
